kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/MetadataCleanup.java

```diff
@@ -0,0 +1,196 @@
+import io.delta.kernel.internal.util.Clock;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MetadataCleanup {
+  private static final Logger logger = LoggerFactory.getLogger(MetadataCleanup.class);
+
+  public MetadataCleanup() {}
+
+  /**
+   * Delete the Delta log files (delta and checkpoint files) that are expired according the table
+   * metadata retention settings. While deleting the log files, it makes sure the time travel
+   * continues to work for all unexpired table versions.
+   *
+   * <p>Here is algorithm:
+   *
+   * <ul>
+   *   <li>Initial the potential delete file list: `potentialFilesToDelete` as empty list
+   *   <li>Initialize the last seen checkpoint file list: `lastSeenCheckpointFiles`. There could be
+   *       one or more checkpoint files for a given version.
+   *   <li>List the delta log files starting with prefix "00000000000000000000." (%020d). For each
+   *       file:
+   *       <ul>
+   *         <li>Step 1: Check if the `lastSeenCheckpointFiles` contains a complete checkpoint, then
+   *             <ul>
+   *               <li>Step 1.1: delete all files in `potentialFilesToDelete`. Now we know there is
+   *                   a checkpoint that contains the compacted Delta log up to the checkpoint
+   *                   version and all commit/checkpoint files before this checkpoint version are
+   *                   not needed.
+   *               <li>Step 1.2: add `lastCheckpointFiles` to `potentialFileStoDelete` list. This
+   *                   checkpoint is potential candidate to delete later if we find another
+   *                   checkpoint
+   *             </ul>
+   *         <li>Step 2: If the timestamp is earlier than the retention period, stop
+   *         <li>Step 3: If the file is a delta log file, add it to the `potentialFilesToDelete`
+   *             list
+   *         <li>Step 4: If the file is a checkpoint file, add it to the `lastSeenCheckpointFiles`
+   *       </ul>
+   * </ul>
+   *
+   * @param engine {@link Engine} instance to delete the expired log files
+   * @param clock {@link Clock} instance to get the current time. Useful in testing to mock the
+   *     current time.
+   * @param tablePath Table location
+   * @param retentionMillis Log file retention period in milliseconds
+   * @throws IOException if an error occurs while deleting the log files
+   */
+  public static void cleanupExpiredLogs(
+      Engine engine, Clock clock, Path tablePath, long retentionMillis) throws IOException {
+    checkArgument(retentionMillis >= 0, "Retention period must be non-negative");
+
+    List<String> potentialLogFilesToDelete = new ArrayList<>();
+    long lastSeenCheckpointVersion = -1;
```

Comment:
you should explain under what scenarios this value will be -1. it is used below to conclude when an incomplete checkpoint is found, it seems

Reply:
indicates we haven't seen any checkpoint yet.